### ISSUE TITLE
Feature request - "kraken2 server"

### ISSUE THREAD

**Issue by UriaMorP** [DerrickWood/kraken2#287]
Hey there,

I'd like to raise an issue previously reported and discussed  in [this issue](https://github.com/DerrickWood/kraken2/issues/233#issue-589955691) as well as [this one](https://github.com/DerrickWood/kraken2/issues/87#issue-406716943) in order to see if my suggestions make any sense before trying to implement them.

Currently, kraken2 doesn't allow processing more than 1 sample at a time.
This means, that for each and every sample, one has to load the database to the memory, or alternatively use the --memory-mapping flag which painfully slows down the execution.

In our lab, we use a [modified version of the GTDB](http://ftp.tue.mpg.de/ebio/projects/struo/GTDB_release89/kraken2/) to process NGS metagenomics samples (hundreds per week).  The problem with this database is that it is about 120GB on disk.

Obviously, in order to process this amount of data, we extensively use HPC services to basically brute force the operation by throwing more computing power on it. However, I get the feeling that working this way is not the most cost-effective (putting gently).

As we all know, once the database is loaded the run-time of a single sample's processing (50M reads) is the short and easy part. This makes me wonder - can we come up with a "load once, process repeatedly" approach that will solve this bottleneck?

My cpp/perl skills are pretty basic, so I might have got it wrong, but what I understand is, that the reference is loaded [here](https://github.com/DerrickWood/kraken2/blob/ed170aab800af5c55ece7eecba7354139d144b88/src/classify.cc#L123-L129) (classify.cc) and assigned to three variables `hash_ptr`,`taxonomy`, `idx_opts`. Then, `ProcessFiles` function is called with these three files (additionally to to read files paths and a taxon call counter). For all I understand, `ProcessFiles` does not modify the three variables (`hash_ptr`,`taxonomy`, `idx_opts`) in a way that is not tractable.  **WHAT IF** we do the following: 

1) Running `main` (classify.cc) until [line 131](https://github.com/DerrickWood/kraken2/blob/ed170aab800af5c55ece7eecba7354139d144b88/src/classify.cc#L130)
2) start an event loop that waits for signals from other processes (or `SIGTERM` once we are done). Now we basically have a server which can handle three types of requests:
    1) SE request, comprised of  `input`, `output`, `report` (paths/PIPES) and `Option opts`
    2) PE request, comprised of  `input` x 2, `output`, `report` (paths/PIPES) and `Option opts`
    3) Termination request, clearing the memory and shutting down the server
3) On PE/SE request, the loop is calls `ProcessFiles`.

After enthusiastically writing this down, I must say that it is clear to me, that someone would have already write it IF it was a good idea. So there are probably many reasons not to do it and I'd like to understand why, and is there anything else we can do in order to save time&money? 

Thanks

**Comment by UriaMorP** [comment]
Got it - working with memory mapping solves this issue! sorry to bother

**Comment by zztin** [comment]
Hi UriaMorP, 
We are facing a similar problem. Can you elaborate what you mean by using `--memory-mapping` slows down the process, and what you mean by 

> Got it - working with memory mapping solves this issue! sorry to bother
?

I would like to use an flag implemented already instead of solution that is not merged into kraken2 main branch (it's many commit behind...)
 https://github.com/DerrickWood/kraken2/issues/233#issuecomment-629775384

If you have more tips, I would be happy to hear!

**Comment by UriaMorP** [comment]
Hey @zztin ,
To be honest, I really don't remember what I meant when I wrote that 

> ... --memory-mapping flag which painfully slows down the execution


Regarding the later comment:

> Got it - working with memory mapping solves this issue!

I wrote it after figuring out the proper use of the `--memory-mapping` flag. To elaborate: 
This flag notifies Kraken that the index is already loaded to the machine's RAM so that kraken will not attempt to reload it. 
However, note that in order for this to work, the db should indeed be located on the RAM. To do so (on our linux machines) we simply copy the db files (hash, opts and taxo .k2d) files to a directory in /dev/shm (say /dev/shm/kraken_db ).
Then calling kraken with `--memory-mapping` flag still requires that you specify the db location (in our case, `--db /dev/shm/kraken_db` ).

When the reference files are as big as the ones we work with, this simple pre-loading of the data extremely speeds up the mapping process.
